Reuse never show again logic - work in progress

YisraelV opened the pull request:
Solving #72417  

@bpasero this is a work in progress, too early for a thorough code review, but the main api is there. Please let me know if you like the way this is going.

The main 2 points to emphasize:
1. The caller of "prompt" doesn't even know "never show again" was selected. They just call prompt and leave this logic to the service.

2. The "never show again" selections are persisted once when the program shuts down, and under a single key.

YisraelV commented:
@bpasero I'll appreciate your thought on the general approach, so I know if to polish it further and make it ready for code review. Thanks in advance.

bpasero commented:
@YisraelV yeah keep on going

YisraelV commented:
@bpasero this is ready for code review. Thanks!

bpasero commented:
@YisraelV your solution is not preserving existing storage keys if I see correctly? We cannot break users that already made a choice by inventing new storage keys, we have to respect the current ones. 

If we introduce options to control this behaviour, I think we should make them available to both prompt and notify methods because we may have some intent to also never show notifications again. In that case I would think the "Do not shot again" button should be a secondary actions (under the gear icon). Does that make sense?

YisraelV commented:
@bpasero totally missed breaking the current keys thanks!

I see two ways around this:

1. simple way: just store each key provided directly to storage, without wrapping it inside an object.

What worries me about this solution is that it could cause collisions if client code provides too generic keys. If I was using the "never show again" functionallity in my code, and didn't know my key was saved directly to stoeage, I would probably think "fileTooBig" was a specific enough key.

2. Check if the provided key is stored as is (without the wrapping object) The check can be removed after a few versions. 

Please tell me what you think.

Regarding the notify method - will do.

bpasero commented:
I think storing the key "as is" is OK, I mean we have been doing this today as well right? With the chance of collisions too.

YisraelV commented:
@bpasero  that's right but until now the prompt caller was dealing with the storage directly, so there might have been more awareness to the need of very specific keys, whereas now the client code doesn't know the key is stored in storage, and therefore might provide too generic keys.

Anyway if you don't think it's something to worry about I'll carry on.

bpasero commented:
Understood, but I think we can document this and probably call this property "`somethingId`" to make this clear it is an identifier that should be unique.

YisraelV commented:
@bpasero I addressed the issues you pointed out.

Thought it was better to add a second parameter to notify method since the first one is supposed to be passed forward to the model.  

Let me know if anything needs changing.

YisraelV commented:
@bpasero fixed let me know of any other changes.

bpasero commented on src/vs/platform/notification/common/notification.ts:
```diff
@@ -62,6 +62,12 @@ export interface INotification extends INotificationProperties {
+	 * options for showing "never show again" prompt.
+	 * by default shown as secondary action
+	 */
+	neverShowOptions?: INeverShowOptions;
```
Move this into `INotificationProperties` you do not need it both here and in `IPromptOptions`

YisraelV replied:
Where is best to document the different default behaviors?

bpasero replied:
What are the differences?

YisraelV replied:
The default in the prompt method is to show the "never show again" as primary action, while in the notify method the default is secondary (I thought that's what you [suggested](https://github.com/microsoft/vscode/pull/73968#issuecomment-498664661). Or maybe you meant secondary in both?)  

the different behavior is documented separately in each interface.

bpasero replied:
I think it is OK that the option is implemented differently depending on where it is being used.

bpasero commented on src/vs/platform/notification/common/notification.ts:
```diff
@@ -171,6 +177,23 @@ export interface IPromptOptions extends INotificationProperties {
+	* If user picks this option, future calls to 'prompt' with the same prompt id will be ignored.
+	* The id is used to persist the selection to storage
+	*/
+	promptId: string;
```
Suggest to just call this `id`

bpasero commented on src/vs/workbench/services/notification/common/notificationService.ts:
```diff
@@ -3,24 +3,30 @@
 		return this._model;
 	}
 
+	constructor(@IStorageService private readonly _storageService: IStorageService) {
```
I prefer to not use underscores unless there is a public `get` access.

bpasero commented on src/vs/workbench/services/notification/common/notificationService.ts:
```diff
@@ -3,24 +3,30 @@
 
 	_serviceBrand: ServiceIdentifier<INotificationService>;
-
+	private readonly neverShowLabel = nls.localize('neverShowAgain', "Don't Show Again");
```
I prefer to inline messages, not have them as variables.

YisraelV replied:
It's for the sake of reuse in "prompt" and "notify".

bpasero replied:
Yeah don't worry about that, we are good as long as the key used for NLS is the same.

YisraelV replied:
Do you mean to make the key a member or just remove the member variable completely?

bpasero replied:
No simply repeat the same nls.localize call twice, i prefer to read the english label in code where it is being used.

bpasero commented on src/vs/workbench/services/notification/common/notificationService.ts:
```diff
@@ -52,12 +58,64 @@ export class NotificationService extends Disposable implements INotificationServ
+				notification.actions.primary = notification.actions.primary || [];
+				notification.actions.primary = [...notification.actions.primary, neverShowAction];
+			}
+
```
Remove all these empty lines

YisraelV commented:
@bpasero thank you very much for the feedback. Please see my last commit.

bpasero commented on src/vs/platform/notification/common/notification.ts:
```diff
@@ -29,6 +29,11 @@ export interface INotificationProperties {
+	/**
+ * options for showing "never show again" action.
+ */
+	neverShowOptions?: INeverShowOptions;
```
```suggestion
	neverShowAgainOptions?: INeverShowAgainOptions;
```

bpasero commented on src/vs/platform/notification/common/notification.ts:
```diff
@@ -29,6 +29,11 @@ export interface INotificationProperties {
 	silent?: boolean;
+
+	/**
+ * options for showing "never show again" action.
```
Please correct the formatting and be a bit more descriptive

bpasero commented on src/vs/platform/notification/common/notification.ts:
```diff
@@ -173,6 +178,17 @@ export interface IPromptOptions extends INotificationProperties {
 }
 
+
+export interface INeverShowOptions {
```
```suggestion
export interface INeverShowAgainOptions {
```

bpasero commented on src/vs/platform/notification/common/notification.ts:
```diff
@@ -173,6 +178,17 @@ export interface IPromptOptions extends INotificationProperties {
+	* The id is used to persist the selection to storage
+	*/
+	id: string;
+	isSecondary?: boolean;
```
Suggest to add a comment explaining the purpose.

bpasero commented on src/vs/workbench/services/notification/common/notificationService.ts:
```diff
@@ -103,6 +159,14 @@ export class NotificationService extends Disposable implements INotificationServ
 		return handle;
 	}
 
+	private neverShow(id: string) {
```
I suggest to inline this method since it is very simple.

bpasero commented on src/vs/workbench/services/notification/common/notificationService.ts:
```diff
@@ -103,6 +159,14 @@ export class NotificationService extends Disposable implements INotificationServ
+		this.storageService.store(id, true, StorageScope.GLOBAL);
+	}
+
+	private isRejected(id: string): boolean {
```
I suggest to inline this method since it is very simple.

bpasero commented on src/vs/workbench/services/notification/common/notificationService.ts:
```diff
@@ -52,12 +57,63 @@ export class NotificationService extends Disposable implements INotificationServ
+					return Promise.resolve();
+				});
+
+			//default to secondary
```
The default cannot be secondary if the option is called isSecondary and is optional. Also it would change the way we used to do it for the large fie optimization case.

YisraelV replied:
How about: 
appearance: 'primary' | 'secondary';
?

bpasero replied:
@YisraelV I think the name `isSecondary?: boolean;` is fine, we use it for `IPromptChoice` as well but if not provided, it needs to default to being primary.

YisraelV replied:
@bpasero you suggested before the default should be "secondary" in the notify method ([here](https://github.com/microsoft/vscode/pull/73968#issuecomment-498664661)) or did I misunderstand you?

bpasero replied:
@YisraelV I would make it consistent with the prompt choice, not sure what I thought back then.

bpasero commented on src/vs/workbench/services/notification/common/notificationService.ts:
```diff
@@ -52,12 +57,63 @@ export class NotificationService extends Disposable implements INotificationServ
+				notification.neverShowOptions.isSecondary = true;
+			}
+			notification.actions = notification.actions || {};
+			if (notification.neverShowOptions.isSecondary) {
```
I would suggest to add the "Don't show again" as last entry to the secondary actions.

YisraelV replied:
I am sorry I forgot what you said about it in your comment. Will fix.

bpasero commented on src/vs/workbench/services/notification/common/notificationService.ts:
```diff
@@ -52,12 +57,63 @@ export class NotificationService extends Disposable implements INotificationServ
+				notification.actions.secondary = [...notification.actions.secondary, neverShowAction];
+			}
+			else {
+				notification.actions.primary = notification.actions.primary || [];
```
I would suggest to add the "Don't show again" as first entry to the primary actions so that it appears first and the most important button is to the right.

YisraelV commented:
@bpasero please see changes. Thank you.

bpasero commented:
@YisraelV thanks, I will go ahead and merge it in with a few cosmetic changes. Since this is only the first half of the desired changes, how about you continue to adopt the new support around the places where today we still manually add a "Don't show again" button?

I am not sure if we persist this choice in global storage scope everywhere, so we may need to beef up the options a little bit to allow for persisting the choice on the workspace scope too.

YisraelV commented:
@bpasero thanks for your guidance! I'll see if someone wants to take this up and if not I will.
